internal/controller/apiextensions/composite/reconciler.go

```diff
@@ -384,12 +443,25 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
 	}
 
 	cr.SetResourceReferences(refs)
-	if err := r.client.Update(ctx, cr); err != nil {
-		log.Debug(errUpdate, "error", err)
+
+	// Create generic runtime object to apply patches from composed resources to either composite
+	// or composite claim. If a claimRef is specified, the composite claim must be patched rather than
+	// the composite, as the composite claim is the owner of the composite specification.
```

Comment:
This was my initial thought as well, but it was not clear to me how the claim reconciler could differentiate between inputs on the claim which should be propagated to the composite, and patches from the composed to the composite which should be propagated to the claim. Server side apply was the main idea I considered but that has its own drawbacks in-terms of API server version requirements.

Updating the claim if one exists allows the claim to remain the source of truth for parameters.

Reply:
> This was my initial thought as well, but it was not clear to me how the claim reconciler could differentiate between inputs on the claim which should be propagated to the composite, and patches from the composed to the composite which should be propagated to the claim.

Yeah I agree this is tricky, but I'd like to do a bit more investigation to see if it's possible since it would allow us to keep a clean separation of responsibilities (i.e. allow the composite reconciler to remain blissfully unaware of claims).

One naive (but relatively simple) approach might be to assume that `spec` is propagated claim -> composite, but `status` is propagated composite -> claim. I suspect bidirectional patching will be most often used to derive the status of the composite resource from the statuses of its composed resources.

Otherwise (possibly a dumb idea - haven't thought this through) could we just patch from the claim to the composite, then patch from the composite back to the claim? Presumably the first patch would ensure the composite was updated with the claim's desired state. 🤔 

> Server side apply was the main idea I considered but that has its own drawbacks in-terms of API server version requirements.

We might consider making the cut if it would help - we had been holding off due to the API server version requirements but if I recall correctly we might already have those requirements due to having migrated to v1 CRDs (iirc both v1 CRDs and SSA require v1.16).

Reply:
> One naive (but relatively simple) approach might be to assume that spec is propagated claim -> composite, but status is propagated composite -> claim. I suspect bidirectional patching will be most often used to derive the status of the composite resource from the statuses of its composed resources.

I created https://github.com/crossplane/crossplane/issues/2081 to track that currently status fields in XRD's are overridden with the default claim/composite status. 

Although that approach would be simpler, propagating back values which were transformed in compositions or values which may have different defaults per composition (rather than the default in XRD) back into the spec seems valuable.

> Otherwise could we just patch from the claim to the composite, then patch from the composite back to the claim? Presumably the first patch would ensure the composite was updated with the claim's desired state. 

If some field's initial value is a zero value, and receives a value from a patch from the composed to the composite, the claim reconciler could patch this value back. But now future reconciles on the claim controller can't differentiate between spec drift between fields on the claim -> composite or patches which need to be propagated back to the claim.

I think this approach could work with server-side apply, my thinking is the composite controller does a server-side apply with force for changes on itself, the claim controller can create a patch diff between the claim and the composite and apply without force this diff on the composite. This would allow spec changes which are not affected by bidirectional patches to persist to the composite. Then the claim is updated based on the results from patching the composite which would be a merge of claim spec and bi-directional patch.

Not sure if that works or makes sense 😄 , which is why my initial draft implementation avoided this and just had the composite reconciler patch the claim instead if it existed.

Reply:
> If some field's initial value is a zero value, and receives a value from a patch from the composed to the composite, the claim reconciler could patch this value back. But now future reconciles on the claim controller can't differentiate between spec drift between fields on the claim -> composite or patches which need to be propagated back to the claim.

Hrm, this is making me think that we might want to treat status as a special case.

In general we follow the [late initialization](https://github.com/kubernetes/community/blob/1d2f05/contributors/devel/sig-architecture/api-conventions.md#late-initialization) convention across our managed resources - i.e. the managed resource is always the source of truth, but we will update a spec field if it is set in the external API (i.e. the cloud provider) and not set in the managed resource. Once the field is set on the managed resource it becomes the source of truth and we never touch it.

I think it makes sense for us to do something similar with XRs - if we only allow composed resources to patch to fields that are not currently set we ensure the XR (and thus claim) are always the source of truth. This pattern would fall apart if applied to status fields though - the source of truth is flipped there and thus you typically want the composed values to overwrite whatever is at the XR/claim level.

I'm wondering if this behaviour should be configurable at the XRD level? i.e. Perhaps we allow only late initialization by default, but you can supply an array of field paths (which would default to only `status`) under which we assume propagation should _always_ happen in reverse - i.e. `composed -> XR -> claim` - even when doing so would overwrite existing values.

Reply:
Hmm yeah late initialization is an interesting way to consider this. If the composite could late initialize spec values with patches from composed, and always own status fields this would solve most of our use-cases around bidirectional patches.

This also solves some of the issues I've been thinking about here where cycles could occur between spec input on the claim and a composition patch which overrode that same field. Without a clear source of truth the claim controller and composite controller would just keep patching that field back and forth.